pkg/controller/master/nodedrain/nodedrain_controller.go

```diff
@@ -77,121 +78,134 @@ func (ndc *ControllerHandler) OnNodeChange(_ string, node *corev1.Node) (*corev1
 	if node == nil || node.DeletionTimestamp != nil {
 		return node, nil
 	}
-	_, ok := node.Annotations[drainhelper.DrainAnnotation]
 
+	_, ok := node.Annotations[drainhelper.DrainAnnotation]
 	if !ok {
 		return node, nil
 	}
 
 	_, forced := node.Annotations[drainhelper.ForcedDrain]
 
-	if ok {
-		// still running a check in the background to avoid maintenance issues when using object annotations
-		// directly
-		err := drainhelper.DrainPossible(ndc.nodeCache, node)
-		if err != nil {
-			return node, err
+	// still running a check in the background to avoid maintenance issues when using object annotations
+	// directly
+	drainPossible, err := drainhelper.DrainPossible(ndc.nodeCache, node)
+	if err != nil {
+		if !drainPossible {
+			// Fail hard, cannot drain this node.
```

Comment:
when error is returned, the controller will retry

need a way to express the `failed to drain` and stop;  e.g. remove the `DrainAnnotation` and put `DrainFailedAnnotation`

but it might belong to another refactor PR

Reply:
I've opened https://github.com/harvester/harvester/issues/8985 for that.

Reply:
Will be fixed in this PR as well.